File: Besties/ListasCirculares/lista.cpp
#include "lista.h"
#include <iostream>
#include <algorithm>
using namespace std;

template <typename T>
Lista<T>::Lista() : cabeza(nullptr), cola(nullptr) {}

template <typename T>
Lista<T>::~Lista() {
    if (!cabeza) return;

    // Romper el enlace circular
    cola->siguiente = nullptr;
    
    while (cabeza != nullptr) {
        Nodo<T>* temp = cabeza;
        cabeza = cabeza->siguiente;
        delete temp;
    }
}

template <typename T>
void Lista<T>::insertarPorCabeza(T data) {
    Nodo<T>* nuevo = new Nodo<T>(data);
    
    if (!cabeza) {
        // Primera inserción
        cabeza = cola = nuevo;
        nuevo->siguiente = nuevo;
        nuevo->anterior = nuevo;
    } else {
        // Insertar al principio de la lista circular
        nuevo->siguiente = cabeza;
        nuevo->anterior = cola;
        cabeza->anterior = nuevo;
        cola->siguiente = nuevo;
        cabeza = nuevo;
    }
}

template <typename T>
void Lista<T>::insertarPorCola(T data) {
    Nodo<T>* nuevo = new Nodo<T>(data);
    
    if (!cola) {
        // Primera inserción
        cabeza = cola = nuevo;
        nuevo->siguiente = nuevo;
        nuevo->anterior = nuevo;
    } else {
        // Insertar al final de la lista circular
        nuevo->anterior = cola;
        nuevo->siguiente = cabeza;
        cola->siguiente = nuevo;
        cabeza->anterior = nuevo;
        cola = nuevo;
    }
}


template <typename T>
void Lista<T>::eliminarPorCabeza() {
    if (!cabeza) return;

    if (cabeza == cola) {
        // Último elemento
        delete cabeza;
        cabeza = cola = nullptr;
    } else {
        Nodo<T>* temp = cabeza;
        cabeza = cabeza->siguiente;
        cabeza->anterior = cola;
        cola->siguiente = cabeza;
        delete temp;
    }
}

template <typename T>
void Lista<T>::eliminarPorCedula(string cedula) {
    if (!cabeza) return;

    Nodo<T>* actual = cabeza;
    do {
        if (actual->data.getCedula() == cedula) {
            if (actual == cabeza) {
                eliminarPorCabeza();
            } else if (actual == cola) {
                // Si es el último elemento
                cola = actual->anterior;
                cola->siguiente = cabeza;
                cabeza->anterior = cola;
                delete actual;
            } else {
                // Elemento en medio de la lista
                actual->anterior->siguiente = actual->siguiente;
                actual->siguiente->anterior = actual->anterior;
                delete actual;
            }
            cout << "Persona con cedula " << cedula << " eliminada exitosamente" << endl;
            return;
        }
        actual = actual->siguiente;
    } while (actual != cabeza);
    
    cout << "No se encontro ninguna persona con la cedula " << cedula << endl;
}


template <typename T>
void Lista<T>::mostrarLista() const {
    if (!cabeza) return;

    Nodo<T>* actual = cabeza;
    do {
        cout << actual->data.getCedula() << " "
             << actual->data.getApellido() << " "
             << actual->data.getNombre() << endl;
        actual = actual->siguiente;
    } while (actual != cabeza);
}

template <typename T>
void Lista<T>::mostrarListaInversa() const {
    if (!cola) return;

    Nodo<T>* actual = cola;
    do {
        cout << actual->data.getCedula() << " "
             << actual->data.getApellido() << " "
             << actual->data.getNombre() << endl;
        actual = actual->anterior;
    } while (actual != cola);
}

template <typename T>
void Lista<T>::eliminarCaracter(char c) {
    if (!cabeza) return;

    Nodo<T>* actual = cabeza;
    do {
        string cedula = actual->data.getCedula();
        cedula.erase(remove(cedula.begin(), cedula.end(), c), cedula.end());
        actual->data.setCedula(cedula);

        string apellido = actual->data.getApellido();
        apellido.erase(remove(apellido.begin(), apellido.end(), c), apellido.end());
        actual->data.setApellido(apellido);

        string nombre = actual->data.getNombre();
        nombre.erase(remove(nombre.begin(), nombre.end(), c), nombre.end());
        actual->data.setNombre(nombre);

        actual = actual->siguiente;
    } while (actual != cabeza);
}


template <typename T>
void Lista<T>::reemplazarCaracter(char original, char reemplazo) {
    if (!cabeza) return;

    Nodo<T>* actual = cabeza;
    do {
        string cedula = actual->data.getCedula();
        replace(cedula.begin(), cedula.end(), original, reemplazo);
        actual->data.setCedula(cedula);

        string apellido = actual->data.getApellido();
        replace(apellido.begin(), apellido.end(), original, reemplazo);
        actual->data.setApellido(apellido);

        string nombre = actual->data.getNombre();
        replace(nombre.begin(), nombre.end(), original, reemplazo);
        actual->data.setNombre(nombre);

        actual = actual->siguiente;
    } while (actual != cabeza);
}

template <typename T>
void Lista<T>::buscarPorCedula(string cedula) {
    if (!cabeza) return;

    Nodo<T>* actual = cabeza;
    do {
        if (actual->data.getCedula() == cedula) {
            cout << "Persona encontrada:" << endl;
            cout << "Cedula: " << actual->data.getCedula() << endl;
            cout << "Nombre: " << actual->data.getNombre() << endl;
            cout << "Apellido: " << actual->data.getApellido() << endl;
            return;
        }
        actual = actual->siguiente;
    } while (actual != cabeza);
    
    cout << "No se encontro ninguna persona con la cedula " << cedula << endl;
}

template <typename T>
Lista<T>::Lista(const Lista& otra) : cabeza(nullptr), cola(nullptr) {
    copiarLista(otra.cabeza);
}

template <typename T>
Lista<T>& Lista<T>::operator=(const Lista& otra) {
    if (this != &otra) {
        while (cabeza != nullptr) {
            eliminarPorCabeza();
        }
        copiarLista(otra.cabeza);
    }
    return *this;
}

template <typename T>
void Lista<T>::copiarLista(const Nodo<T>* otraCabeza) {
    if (!otraCabeza) return;

    const Nodo<T>* actual = otraCabeza;
    do {
        insertarPorCola(actual->data);
        actual = actual->siguiente;
    } while (actual != otraCabeza);
}

File: TrabajosSaray/listaspython/listas_circulares/lista.py
import sys
import os

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
from nodo import Nodo
from correo import GeneradorCorreo
from persona import Persona

class Lista:
    def __init__(self):
        self.cabeza = None
        self.cola = None
        self.generador_correo = GeneradorCorreo()
        self.cedulas_registradas = set() 

    def validar_cedula_duplicada(self, cedula):
        return cedula not in self.cedulas_registradas

    def insertar_por_cabeza(self, persona):
        if not self.validar_cedula_duplicada(persona.cedula):
            print(f"\nError: La cédula {persona.cedula} ya está registrada en el sistema.")
            return False
            
        nuevo = Nodo(persona)
        if not self.cabeza:
            self.cabeza = self.cola = nuevo
            nuevo.siguiente = nuevo
            nuevo.anterior = nuevo
        else:
            nuevo.siguiente = self.cabeza
            nuevo.anterior = self.cola
            self.cabeza.anterior = nuevo
            self.cola.siguiente = nuevo
            self.cabeza = nuevo
        
        self.cedulas_registradas.add(persona.cedula)
        return True

    def insertar_por_cola(self, persona):
        if not self.validar_cedula_duplicada(persona.cedula):
            print(f"\nError: La cédula {persona.cedula} ya está registrada en el sistema.")
            return False
            
        nuevo = Nodo(persona)
        if not self.cola:
            self.cabeza = self.cola = nuevo
            nuevo.siguiente = nuevo
            nuevo.anterior = nuevo
        else:
            nuevo.anterior = self.cola
            nuevo.siguiente = self.cabeza
            self.cola.siguiente = nuevo
            self.cabeza.anterior = nuevo
            self.cola = nuevo
        
        self.cedulas_registradas.add(persona.cedula)
        return True

    def eliminar_por_cedula(self, cedula):
        if not self.cabeza:
            return False

        actual = self.cabeza
        while True:
            if actual.data.cedula == cedula:
                if actual == self.cabeza:
                    if self.cabeza == self.cola:
                        self.cabeza = self.cola = None
                    else:
                        self.cabeza = actual.siguiente
                        self.cabeza.anterior = self.cola
                        self.cola.siguiente = self.cabeza
                elif actual == self.cola:
                    self.cola = actual.anterior
                    self.cola.siguiente = self.cabeza
                    self.cabeza.anterior = self.cola
                else:
                    actual.anterior.siguiente = actual.siguiente
                    actual.siguiente.anterior = actual.anterior
                
                self.cedulas_registradas.remove(cedula)
                print(f"\nPersona con cédula {cedula} eliminada exitosamente")
                return True
            
            actual = actual.siguiente
            if actual == self.cabeza:
                break
        
        print(f"\nNo se encontró ninguna persona con la cédula {cedula}")
        return False

    def mostrar_lista(self):
        if not self.cabeza:
            print("\nLista vacía")
            return

        actual = self.cabeza
        while True:
            print(f"Cédula:   {actual.data.cedula}")
            print(f"Nombre:    {actual.data.nombre}")
            print(f"Apellido:  {actual.data.apellido}")
            print(f"Correo:    {actual.data.correo}")
            print("-" * 40)
            
            actual = actual.siguiente
            if actual == self.cabeza:
                break

    def copiar_lista(lista_origen, lista_destino):
       
        lista_destino.cabeza = None
        lista_destino.cola = None
        lista_destino.cedulas_registradas.clear()
        
        if not lista_origen.cabeza:
            return
            
        actual = lista_origen.cabeza
        while True:
            nueva_persona = Persona(
                actual.data.cedula,
                actual.data.apellido,
                actual.data.nombre,
                actual.data.correo
            )
            lista_destino.insertar_por_cola(nueva_persona)
            actual = actual.siguiente
            if actual == lista_origen.cabeza:
                break

    def eliminar_caracter(self, c):
        if not self.cabeza:
            return

        actual = self.cabeza
        while True:
            actual.data.cedula = actual.data.cedula.replace(c, '')
            actual.data.nombre = actual.data.nombre.replace(c, '')
            actual.data.apellido = actual.data.apellido.replace(c, '')
            actual.data.correo = self.generador_correo.generar_correo(
                actual.data.nombre, 
                actual.data.apellido
            )
            
            actual = actual.siguiente
            if actual == self.cabeza:
                break

    def reemplazar_caracter(self, original, reemplazo):
        if not self.cabeza:
            return

        actual = self.cabeza
        while True:
            actual.data.cedula = actual.data.cedula.replace(original, reemplazo)
            actual.data.nombre = actual.data.nombre.replace(original, reemplazo)
            actual.data.apellido = actual.data.apellido.replace(original, reemplazo)
            actual.data.correo = self.generador_correo.generar_correo(
                actual.data.nombre, 
                actual.data.apellido
            )
            
            actual = actual.siguiente
            if actual == self.cabeza:
                break

    def buscar_por_cedula(self, cedula):
        if not self.cabeza:
            return None
            
        actual = self.cabeza
        while True:
            if actual.data.cedula == cedula:
                return actual.data
            actual = actual.siguiente
            if actual == self.cabeza:
                break
        return None
    
    def limpiar_pantalla():
        import os
        os.system('cls' if os.name == 'nt' else 'clear')